src/modules/dual_path.py

```diff
@@ -482,3 +496,141 @@ def evaluate_network(self, set_names, top_down_language_activation=False):
+                               weights_role_concept[line_idx],
+                               weights_concept_role[line_idx],
+                               weights_role_identif[line_idx],
+                               weights_identif_role[line_idx],
+                               event_semantics[line_idx],
+                               backpropagate=True, activate_target_lang=True)
+                
+            # Process target as if testing
+            else:
+                target_pos, target_sentence_idx, target_lang = line.target_pos, line.target_sentence_idx, line.lang
+                alt_sentence_idx, alternative_sentence = line.alt_sentence_idx, line.alt_sentence
+                produced_idx, entropy_idx = self.feed_line(line.target_sentence_idx + buffer_to_produce,
+                                                       target_lang_act[line_idx],
+                                                       weights_role_concept[line_idx],
+                                                       weights_concept_role[line_idx],
+                                                       weights_role_identif[line_idx],
+                                                       weights_identif_role[line_idx],
+                                                       event_semantics[line_idx],
+                                                       activate_target_lang=not top_down_language_activation)
+
+                produced_sentence = self.inputs.sentence_from_indices(produced_idx)
+                produced_pos = self.inputs.sentence_pos(produced_idx)
+                alt_pos = self.inputs.sentence_pos(alt_sentence_idx)
+
+                # We ignore periods. Quite a few sentences without periods are produced that are 
+                # otherwise correct
+                period_idx = self.inputs.df_query_to_idx("pos == '.'")
+                produced_idx_no_period = list(filter(lambda i: i not in period_idx, produced_idx))
+                target_sentence_idx_no_period = list(filter(lambda i: i not in period_idx, target_sentence_idx))
+                alt_sentence_idx_no_period = list(filter(lambda i: i not in period_idx, alt_sentence_idx))
+                    
+                target_pos_no_period = self.inputs.sentence_pos(target_sentence_idx_no_period)
+                alt_pos_no_period = self.inputs.sentence_pos(alt_sentence_idx_no_period)
+                produced_pos_no_period = self.inputs.sentence_pos(produced_idx_no_period)
+
+                has_correct_pos, has_wrong_det, has_wrong_tense, correct_meaning, cs_type = (False, False, False,
+                                                                                             False, None)
+                correct_alternative_meaning, alt_has_wrong_det, alt_has_wrong_tense = (False, False, False)
+
+                # check for zero length to prevent list index error in is_sentence_gramatical_or_flex()
+                if len(produced_pos_no_period) == 0:
```

Comment:
(Another FYI) I'd simplify it by intializing "is_grammatical, alt_is_grammatical = False, False" and only proceed "if produced_pos_no_period:"    (because it's the same as checking if len(produced_pos_no_period) > 0 and it skips the extra computation of determining the length.